rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -54,329 +55,243 @@ use super::{
+            metrics,
+            rx: rx_prepare,
+        } = self;
+        let prepare_queue: OpQueue = Default::default();
+        let confirm_queue: OpQueue = Default::default();
+
+        let (tx_submit, rx_submit) = mpsc::channel(1);
+
+        let tasks = [
+            spawn(receive_task(domain, rx_prepare, prepare_queue.clone())),
+            spawn(prepare_task(
+                tx_submit,
+                prepare_queue.clone(),
+                metrics.clone(),
+            )),
+            spawn(submit_task(
+                rx_submit,
+                prepare_queue.clone(),
+                confirm_queue.clone(),
+                metrics.clone(),
+            )),
+            spawn(confirm_task(prepare_queue, confirm_queue, metrics)),
+        ];
+
+        for i in try_join_all(tasks).await? {
+            i?
         }
+        Ok(())
     }
+}
 
-    /// Tick represents a single round of scheduling wherein we will process
-    /// each queue and await at most one message submission. It is extracted
-    /// from the main loop to allow for testing the state of the scheduler
-    /// at particular points without having to worry about concurrent
-    /// access.
-    async fn tick(&mut self) -> Result<()> {
-        // Pull any messages sent by processor over channel.
-        loop {
-            match self.rx.try_recv() {
-                Ok(msg) => {
-                    self.run_queue.push(Reverse(msg));
-                }
-                Err(TryRecvError::Empty) => {
-                    break;
-                }
-                Err(_) => {
-                    bail!("Disconnected rcvq or fatal err");
-                }
-            }
-        }
-
-        // TODO: Scan verification queue, dropping messages that have been
-        // confirmed processed by the mailbox indexer observing it. For any
-        // still-unverified messages that have been in the verification queue
-        // for > threshold_time, move them back to the wait queue for further
-        // processing.
-
-        self.metrics
-            .run_queue_length_gauge
-            .set(self.run_queue.len() as i64);
-
-        // check if the next message is going to be processable
-        if let Some(Reverse(PendingMessage {
-            next_attempt_after: Some(retry_after),
-            ..
-        })) = self.run_queue.peek()
-        {
-            if Instant::now() < *retry_after {
-                return Ok(());
-            }
-        }
+#[instrument(skip_all)]
+async fn receive_task(
+    domain: HyperlaneDomain,
+    mut rx: mpsc::UnboundedReceiver<Box<DynPendingOperation>>,
+    prepare_queue: OpQueue,
+) -> Result<()> {
+    // Pull any messages sent to this submitter
+    while let Some(op) = rx.recv().await {
+        trace!(?op, "Received new operation");
+        // make sure things are getting wired up correctly; if this works in testing it
+        // should also be valid in production.
+        debug_assert_eq!(*op.domain(), domain);
```

Comment:
Maybe we should use this macro for our PendingMessage submission/prepare data checks?

Reply:
I do not follow

Reply:
Looks like you agree:
https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2195#discussion_r1193916648

Reply:
Ah, I see, yes, added